pkg/sentry/syscalls/linux/vfs2/mq.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2021 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vfs2
+
+import (
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/sentry/arch"
+	"gvisor.dev/gvisor/pkg/sentry/kernel"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/mq"
+)
+
+// MqOpen implements mq_open(2).
+func MqOpen(t *kernel.Task, args arch.SyscallArguments) (uintptr, *kernel.SyscallControl, error) {
+	nameAddr := args[0].Pointer()
+	flag := args[1].Int()
+	mode := args[2].ModeT()
+	attrAddr := args[3].Pointer()
+
+	name, err := t.CopyInString(nameAddr, mq.MaxName)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	rOnly := flag&linux.O_RDONLY == linux.O_RDONLY
+	wOnly := flag&linux.O_WRONLY == linux.O_WRONLY
+	readWrite := flag&linux.O_RDWR == linux.O_RDWR
+
+	create := flag&linux.O_CREAT == linux.O_CREAT
+	exclusive := flag&linux.O_EXCL == linux.O_EXCL
+	block := flag&linux.O_NONBLOCK != linux.O_NONBLOCK
+
+	var attr linux.MqAttr
+	var attrPtr *linux.MqAttr
+	if attrAddr != 0 {
+		if _, err := attr.CopyIn(t, attrAddr); err != nil {
+			return 0, nil, err
+		}
+		attrPtr = &attr
+	}
+
+	r := t.IPCNamespace().PosixQueues()
+	queue, err := r.FindOrCreate(t, name, rOnly, wOnly, readWrite, create, exclusive, block, statusFlags(flag, rOnly, wOnly, readWrite), linux.FileMode(mode), attrPtr)
```

Comment:
This is a lot of booleans to pass to a function.  Can we define a struct type to hold all these options?  And since only one of {rOnly, wOnly, readWrite} can be true, it seems like a good fit for an enum.
```
type AccessType int
const (
  ReadOnly AccessType = iota
  WriteOnly
  ReadWrite
}

type MqOpts struct {
  Name string
  AccessType AccessType
  Create bool
  Exclusive bool
  Block bool
} 
```
This same argument type could be passed to the Impls too.

Reply:
Also, why do we need to pass `statusFlags(...)` since that information seems redundant with the data we are already passing in.

Reply:
Done. I updated `FindOrCreate` to use `MqOpts` and create the status flag, and updated all applicable functions to use `AccessType` instead of `rOnly, wOnly, readWrite bool`.